Scheduler and Deck info through ContentProvider

wlky opened the pull request:
In order to allow reviews from other apps and devices (especially Android Wear Smartwatches) I extended the CardContentProvider and FlashCardsContract with methods and classes that allow external apps to:

-request a list containing the name and id of the users decks
-request the next scheduled card
-request which deck is selected right now
-update the ease of a card (answer the card)
-update the selected deck (choose a deck)

timrae commented:
Hi thanks for this, since it's quite an important class we'll need to go through and review it line by line... are you ok with that? To start with I think it needs a bit more documentation, and some unit tests to check that it's working and continues to work in the future through libanki changes. See #725 as a reference

federvieh commented:
@wlky I really, really like your idea! You can find the existing test cases in com.ichi2.anki.tests/ContentProviderTest.java

federvieh commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/FlashCardsContract.java:
```diff
@@ -840,4 +838,91 @@ private FlashCardsContract() {
         public static final String CONTENT_TYPE = "vnd.android.cursor.dir/vnd.com.ichi2.anki.card";
     }
+
+    public static class ReviewInfo {
```
Would be good if you could add a description of how you intend this class (and also the "Deck" class) to be used. Please describe what its purpose is, but also how third party apps should use this class. See the other classes for reference.

wlky commented:
glad you like it :) I've been using the modified contentprovider in combination with my [AnkiDroid-Wear](https://github.com/wlky/AnkiDroid-Wear) app to learn japanese on my watch for over a week now, and it has become on the the main things I use my watch for (next to seeing what time it is, and reading notifications :P)

I added some documentation for the ReviewInfo and the Deck classes.
I'll take a look at the unit tests and documentation of the CardContentProvider changes next.

wlky commented:
The URI "CONTENT_SELECTED_URI" is meant to be used to access, which deck is currently selected by the user for reviewing. This allows external apps to show which deck is currently being reviewed, and also, via the update method, to change the selected deck.

Specific use-case:
In the android Wear app I want to be able to show a list of all decks, indicate which deck is currently selected for reviewing, and allow the user to change the selected deck.

federvieh commented:
Okay, I see your point. But this means, that setting the “selected desk”
alters the return value of another URI. Wouldn't it make more sense to make
the deck a parameter of the “due cards” URI (either by expanding the URI or
by using it in the selection parameter).

This  again leads to another, more fundamental question: Is our cards URI
flawed? What's the difference between the “cards” and the “due cards”
really? Couldn't “is due” be a parameter when you query for cards? I know
that right now you can only query cards in the form
“…/notes/[noteID]/cards”. But should this be changed, so that you can query
for cards by “…/cards”, i.e. without giving the note ID in the URI?

wlky commented:
Hmm right, passing the deck id as a parameter when requesting the due cards does make sense.

But I don't think "is due" should be a parameter, because there are additional infos when reviewing a card (like the number of buttons to show) that you don't need for regular cards. But in the way it is right now, with the ReviewInfo, you can request the ReviewInfo that contains this addinional information and also the note id and card ord, with which you can request the orinial card if you need to (so kind of like the RequestTable being a different table with a foreign key based on the unique combination of note id, card ord).

federvieh commented:
Allright, then let's forget about merging “cards” and “due cards” URIs.
It's probably more trouble then benefit.

timrae commented:
"schedule" could be a better name instead of "due_cards"

On Tue, May 5, 2015 at 1:15 AM, Frank Oltmanns <notifications@github.com>
wrote:

> Allright, then let's forget about merging “cards” and “due cards” URIs.
> It's probably more trouble then benefit.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ankidroid/Anki-Android/pull/823#issuecomment-98770026>
> .
>

wlky commented:
In the latest commit I removed the selected deck from the update method, and added a parameter to the due cards query that decides which deck to get the scheduled cards from. I also added another parameter to limit the number of scheduled cards to query for.

Regarding a different topic:
Is there a way to get the media resources (most importantly sounds and images) of a flashcard?

federvieh commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -24,17 +24,26 @@
 import android.database.MatrixCursor;
 import android.database.SQLException;
 import android.net.Uri;
+import android.util.Log;
```
I think, you are now using Timber everywhere in your code, so this import is not needed.

federvieh commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -85,6 +97,10 @@
         sUriMatcher.addURI(FlashCardsContract.AUTHORITY, "notes/#/cards/#", NOTES_ID_CARDS_ORD);
         sUriMatcher.addURI(FlashCardsContract.AUTHORITY, "models", MODELS);
         sUriMatcher.addURI(FlashCardsContract.AUTHORITY, "models/#", MODELS_ID);
+        sUriMatcher.addURI(FlashCardsContract.AUTHORITY, "due_cards/", SCHEDULE);
```
I think, what @timrae meant was that also the URI should be called "scheduled".

federvieh commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -70,6 +79,9 @@
     private static final int MODELS_ID = 2001;
+    private static final int SCHEDULE = 3000;
+    private static final int DECKS = 4000;
+    private static final int DECK_SELECTED = 4001;
```
Is this still required?

wlky replied:
I think so, they are necessary for the UriMatcher to work. In Case you meant if it is still required to have all three of them: the DECKS constant is for requesting the deck list, the SCHEDULE for requesting the scheduled cards. DECK_SELECTED could also be replaced by adding a parameter to the DECK_LIST (isSelected or something)

federvieh commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/FlashCardsContract.java:
```diff
@@ -840,4 +837,264 @@ private FlashCardsContract() {
+     * <tr>
+     * <td>long</td>
+     * <td>deckID</td>
+     * <td>Last selected deck</td>
```
As far as I understand the code in CardContentProvider, you don't use the deck from the last call to this URI, but the deck that was last selected by the user inside the AnkiDroid UI, correct? If so, this should be explained here.

federvieh commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/FlashCardsContract.java:
```diff
@@ -840,4 +837,264 @@ private FlashCardsContract() {
+                NEXT_REVIEW_TIMES,
+                QUESTION_SIMPLE,
+                ANSWER_SIMPLE,
+                EASE
```
Can EASE really be part of the default projection for queries? According to the description above, EASE is write-only.

wlky replied:
Oh you're right, that doesn't make sense

federvieh commented on AnkiDroid/src/main/java/com/ichi2/libanki/Card.java:
```diff
@@ -409,6 +409,15 @@ public String getPureAnswerForReading() {
+            return s;
+        }
+        return s.substring(pos + target.length()).trim();
+    }
```
This should be reviewed by a senior member.

wlky replied:
ehm to be honest I think it was the wrong approach to put this in this pull request anyway. 
getPureAnswerForReading is already supposed to do what I want, but it is, for some reason, limited to when there are two new lines after the <hr id=answer> which is not always the case. 

So I think this is a separate bug. Should I create a new pull request for that?

hssm replied:
The default template uses two new lines so that's probably what the original developer had in mind. Obviously this is still error-prone but it should cover most cases.

A new pull request would be clearer, so I would prefer that if you could. Since it's no longer only related to TTS you can rename it to what you have here and add a comment above it to say what it does.

federvieh commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -392,6 +499,30 @@ public int update(Uri uri, ContentValues values, String selection,
+                if (cardOrd != -1 && noteID != -1) {
+                    Card cardToAnswer = col.getNote(noteID).cards().get(cardOrd);
+                    answerCard(col, col.getSched(), cardToAnswer, ease);
+                }
```
This could potentially cause a NPE if an invalid note ID or card ordinal are given.

wlky replied:
Right that's not good, and it's also kind of a duplicate of what getCardFromUri does internally.
What do you think about having a new Method getCard(long noteId, int ord) that is called from my code an from getCardFromUri?

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/FlashCardsContract.java:
```diff
@@ -792,7 +789,7 @@ private FlashCardsContract() {
     public static class Card {
         /**
-         * This is the ID of the note that this card belongs to (i.e. {@link Note#_ID}).
+         * This is the ID of the note that this card belongs to (i.e. {@link FlashCardsContract.Note#_ID}).
```
Are all these `FlashCardsContract.` prefixes that you've added necessary in the javadoc?

wlky replied:
they didn't hurt, but weren't necessary either. I'll remove them in the next commit

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -291,6 +311,92 @@ public Cursor query(Uri uri, String[] projection, String selection,
+                                if (col.getDecks().get(deckIdOfTemporarilySelectedDeck, false) != null) {
+                                    col.getDecks().select(deckIdOfTemporarilySelectedDeck);
+                                } else {
+                                    Timber.i("Requested deck with id " + deckIdOfTemporarilySelectedDeck + " is was not found in deck list");
```
Please use formatted strings with Timber:
`Timber.i("Requested deck with id %d  was not found in deck list", deckIdOfTemporarilySelectedDeck);`

timrae replied:
Also it should be a warning not info

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -291,6 +311,92 @@ public Cursor query(Uri uri, String[] projection, String selection,
+                int selectionArgIndex = 0;
+
+                if (selection != null) {
+                    String[] args = selection.split(","); //split selection to get arguments like "limit=?"
```
Is having the selected deck as an argument any better / worse than having it as part of the URL itself? E.g.
`schedule\deckid\...`

wlky replied:
I don't think it makes any difference. Doing it this way felt more natural from a database point of view since the deckID is a selecting argument on the Deck database table

timrae commented:
I can't really see a usage case for read-only access to the `DECK_SELECTED` URI... It could be useful if it were necessary to sync the selected deck between different clients which are sharing the AnkiDroid provider, but that is not possible if it's read-only, so personally I think it should either be read-write or not exist at all (i.e. you'd track the selected deck internally in your app, independent from AnkiDroid's selected deck)

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -291,6 +311,92 @@ public Cursor query(Uri uri, String[] projection, String selection,
+                        for (int i = 0; i < buttonCount; i++) {
+                            buttonTexts.put(col.getSched().nextIvlStr(currentCard, i + 1, true));
+                        }
+                        addReviewInfoToCursor(currentCard, buttonTexts, buttonCount, currentCard.qSimple(), currentCard.getPureAnswer(), rv, col, columns);
```
How about having an optional flag which enables or disables the "simple mode" which you are using in your app? I would guess that most apps that would use this ContentProvider would probably want to get the full Q&A including the CSS, rather than `qSimple()` and `getPureAnswer()`

wlky replied:
you can still get the full Q&A of a card by using the noteID and cardOrd provided in the ReviewInfo and starting a second request. 
If the external app doesn't need the simpleQ and simpleA it can use a projection other than the default projection which doesnt include those columns.
On the other hand your approach with the flag sounds a little bit more user friendly, since it allows getting most of the important info about a scheduled card with one request.

timrae replied:
Sorry I don't understand, what would the second request look like? and yeah I think even if that's the case, it'd be much better if they could get it with one request. The default behavior should be non-simple IMO

wlky replied:
My thought was to reduce the amount of information in ReviewInfo to additional stuff that can not be requested through other requests.
This way you could e.g. get a complete list of scheduled cards' ids and then selectively request the Q&A for only the next scheduled card. For this you would start a second request to the CardContentProvider using the notes/#/cards/# uri and the noteID and cardOrd of the first ReviewInfo element you got from the first request.

What do you think about adding the normal question and answer as additional columns in ReviewInfo, and leaving the simple Q&A out of the default projection? This would allow the external app to request, depending on the projection, no, the simple, the normal or both the normal and the simple Q&A in the same request

timrae replied:
Oh sorry I see now, I had actually forgotten that we had put q & a into the `notes/#/cards` cursor.

> What do you think about adding the normal question and answer as additional columns in ReviewInfo, and leaving the simple Q&A out of the default projection?

That sounds reasonable, but I also agree that it may not make a lot of sense to duplicate the content across the 2 URIs. What do you think about putting simple q and simple a into the notes/#/cards URI? This would have the advantage that one could get the simple q and simple a from e.g. a browser-like Activity irrespective of whether it's available in the schedule, but you would have to do two requests. Sorry for the 180 degree turn!

wlky replied:
Yeah, sounds like a good solution

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -291,6 +311,92 @@ public Cursor query(Uri uri, String[] projection, String selection,
+                }
+
+                col.getSched().reset();
+                for (int k = 0; k< limit; k++){
```
What's the usage case for getting multiple cards at once? I would think that you would normally only want one card at a time, because the schedule can change according to the answers of previous questions, and also for performance reasons (i.e. rendering Q&A is quite an expensive operation)

wlky replied:
in my case (i.e. using Anki on a SmartWatch) it was quite handy to get the next couple of scheduled cards in order to create a buffer. Especially for media files this is quite handy, because if they are note buffered it takes a couple of seconds to transfer them to the watch.

timrae replied:
Thanks, I see!

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -291,6 +311,92 @@ public Cursor query(Uri uri, String[] projection, String selection,
+                ArrayList<JSONObject> allDecks = col.getDecks().all();
+
+                MatrixCursor rv;
+                String[] columns = ((projection != null) ? projection : FlashCardsContract.Deck.DEFAULT_PROJECTION);
```
I think it would be useful to add the deck counts, and possibly more deck properties (options, etc?) to this table. That doesn't need to go in this PR if you're not interested in that, but it would be good if we finalize the format before we release this into the wild in v2.5.

wlky replied:
deck counts like the number of cards per deck?
And the general options per deck like: maxAnswerTime, showAnswerTimer, autoPlayAudio and replayQuestion?

timrae replied:
deck counts meaning the 3 numbers that are shown next to each deck in the
deck picker... although that's strictly part of the scheduler.
and options yes! probably it would be fine to simply provide the JSONarray
verbatim

On Tue, Jun 2, 2015 at 9:45 PM, Yannik <notifications@github.com> wrote:

> In
> AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java
> <https://github.com/ankidroid/Anki-Android/pull/823#discussion_r31518070>:
>
> > +                    }else{
> > +                        break;
> > +                    }
> > +                }
> > +
> > +                if (deckIdOfTemporarilySelectedDeck != -1) {//if the selected deck was changed
> > +                    col.getDecks().select(selectedDeckBeforeQuery); //change the selected deck back to the one it was before the query
> > +                }
> > +                return rv;
> > +            }
> > +            case DECKS: {
> > +
> > +                ArrayList<JSONObject> allDecks = col.getDecks().all();
> > +
> > +                MatrixCursor rv;
> > +                String[] columns = ((projection != null) ? projection : FlashCardsContract.Deck.DEFAULT_PROJECTION);
>
> deck counts like the number of cards per deck?
> And the general options per deck like: maxAnswerTime, showAnswerTimer,
> autoPlayAudio and replayQuestion?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ankidroid/Anki-Android/pull/823/files#r31518070>.
>

timrae commented:
Thanks again for this, it's a really nice addition to AnkiDroid! Sorry for my slow response, I've gone through it today, and have added a few comments and questions. Once those are addressed, and you add some unit tests, I hope we can get it merged relatively quickly!

wlky commented:
The latest commit should fix most of the comments from above. 

Also I added a way to directly request one single deck (.../decks/#). The implementation is not ideal though, because it still builds the whole DueTree because I couldn't find an easy way to get the deck count for a single deck..

Simple Q&A are now part of the Card class and the Deck class has two more columns.

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java:
```diff
@@ -291,12 +312,119 @@ public Cursor query(Uri uri, String[] projection, String selection,
+                                }
+                            }
+                        } catch (NumberFormatException nfe) {
+                            nfe.printStackTrace();
```
-> `Timber.e(nfe, "some better description about the error and why it could have happened")`

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/FlashCardsContract.java:
```diff
@@ -764,6 +761,21 @@ private FlashCardsContract() {
+     * <td>String</td>
+     * <td>{@link #QUESTION_SIMPLE}</td>
+     * <td>read-only</td>
+     * <td>The question for this card in the simplified form.
```
Please define exactly what "simplified form" means (I think it just means "without the card styling CSS" right?)

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/FlashCardsContract.java:
```diff
@@ -840,4 +862,324 @@ private FlashCardsContract() {
+     * <td>int</td>
+     * <td>limit</td>
+     * <td>1</td>
+     * <td>The maximum number of cards (rows) that will be returned. In case the deck has fewer scheduled cards, the returned number of cards will be lower than the limit.
```
Could you please put newlines where the comments go past the end of the screen on github?

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/provider/FlashCardsContract.java:
```diff
@@ -840,4 +862,324 @@ private FlashCardsContract() {
+     * Requesting a single deck can be done the following way:
+     *
+     * <pre>
+     *     long deckId = 123456 //<-- insert real deck ID here
```
did you forget `<code>` tags?

timrae commented:
Thanks it's looking good! I added a few small comments. 
and still waiting for the unit tests ;)

@federvieh @flerda 
anything to add?

hssm commented:
>I couldn't find an easy way to get the deck count for a single deck..

You can only do it for the current deck with `Sched.counts()` which is what the study options (deck overview) screen does. I wouldn't worry about anyway, it's not a very expensive operation and we do it regularly throughout the application on even the slightest changes.

wlky commented:
All right, I think I addressed & fixed all of your comments @timrae (even the unit tests :P)

When running the unit tests though, I noticed that the testMoveToOtherDeck() keeps failing when there are decks+cards already existing in the tested AnkiDroid instance.
I think I have traced the error down though: The notes created in setUp() are not assigned to a Deck, so AnkiDroid assigns them to the models' default deck, which is where all of the other cards of that model are in (?)
Anyway, the testMoveToOtherDeck() fails when the note is not in the Default deck.
Don't know if it's important, since the tests run smoothly on a "clean" version of AnkiDroid, but I thought I'd let you know.

timrae commented:
Thanks! Yeah I've noticed that happening as well, it would be nice to fix someday but for now it's fine as long as it works on Travis CI. I think it's looking good to go now, would you mind squashing down to a single commit?

wlky commented:
Well, that was a first :D have never squashed my commits before^^
But I think it worked as it was supposed to.
